### DISCUSSION TITLE
Level frames in Function editor and Expressions

### DISCUSSION THREAD

**Discussion by b4zz4** [opentoonz/opentoonz#6547] [Ideas]
**Is your feature request related to a problem? Please describe**
I want a way to manipulate from the "Function editor" the "level draw" sheet, being able to alter the drawings using programming.

**Describe the solution you'd like**
Add another column in the "Function editor", the column "level" for example

![2020-08-08-194519_1066x546_scrot](https://user-images.githubusercontent.com/643833/89721181-c02f0580-d9b0-11ea-99cd-048b49cafade.png)

**Comment by RodneyBaker** [comment]
I think we need more information on what it is you are hoping to accomplish.

Would the 'Level' column simply identify the level targeted in the column?
Would this be controlled by an index number?
Example:  Column 1 contains 5 levels (two of which are toonz raster and three of which are vector).  
Identifying the level with a 1 would target the first level in the column whereas entering a 4 would target the fourth level in the column.    Entering a value of 6 would do nothing as there is not yet a level 6 exposed in the column.

Further..
Will identifying a specific level be enough?
Perhaps entering an integer would target a specific level whereas a float would target a specific frame, group, stroke... in short a specific ID to target.
Example:  Entering 1.4724 would target level 1 (in the current column context of the Function Editor) and a group identified with the number 4724 which might be a number of things depending on what the user had designated with that identification number.

We do need a method to drill down inside a column and to grab a specific 'object' inside a column.
It would be good to know more about what the specific needs are and what success looks like.

I do think we need to be exposing things like Groups in Opentoonz for the user to interact with.
Targeting a specific level within a column is a good start.

**Comment by b4zz4** [comment]
Sheet would be a more appropriate name.
The idea is to be able to associate rotations, and artifacts to create harmonic clipping without having to go back to the timeline and to create automatic variations with drawing changes.

![sheet](https://user-images.githubusercontent.com/643833/89723299-be723b80-d9ca-11ea-88e0-43384e077945.png)

**Comment by b4zz4** [comment]
![current result](https://user-images.githubusercontent.com/643833/89724568-3b0d1600-d9db-11ea-9aa0-09048e080716.gif)  ![rotation link to frames x-sheet](https://user-images.githubusercontent.com/643833/89724570-3e080680-d9db-11ea-8124-bdf7dd114c70.gif)

**Comment by RodneyBaker** [comment]
I follow what you are after but am mostly trying to see how we might get from here (where we currently are) to there.
An initial step might to provide a reference to the level and frame in the Function Editor.
Your mockup of course adds a new functionality to the Edit/Animate Tool that currently isn't there that would be useful; mouse tracking.  Currently we'd have to use on onscreen controller to achieve that particular functionality (we can currently supply a Null column, drawing or even FX manipulator to achieve that aspect via controller).

**Comment by b4zz4** [comment]
You don't have to follow the mouse.
Just associate expression programming with sheet (sheet x) and Function editor
I know it's difficult, but it helps to create very ductile characters in the cutouts.

![2020-08-09-023850_319x165_scrot](https://user-images.githubusercontent.com/643833/89725779-7b739080-d9e9-11ea-9cca-39b45cbdf440.png)

And import file, automatic interpolation, etc

**Comment by beeheemooth** [comment]
I had the same idea! Having the ability to bind the change of the number of one column of the frame to Xsheet with the key parameters (rotate, scale ... e.t.c) of another column (or the same, but avoiding the "circular reference"), we can make an imitation of smartbones or bone morphing.
However, if you think about it, you can predict many questions that are associated with this.
1. If the replacement of a drawing depends on a change in the key interpolation, then a change in one key will lead to a change in all subsequent drawings
2. How should this feature behave if the user decides to add or remove keys, move them or change interpolation?
3. How should this feature behave in columns with Plastic meshes in cases when the skeleton number changes?
4. What should be the result of the "explosion" of the sub-xsheet?
5. Should the effect keys affect the frame number change?
6. Should there be a limitation for "not yet existing" frames?
7 ...
 But the idea is really interesting, and it looks logical for Opentoonz.

**Comment by b4zz4** [comment]
Sure, you must follow the logic of the opentoonz. 
I was thinking of looking for the variable that contains the list of drawings and associating it with the Function Editor at least globally. In other words, if you associate the x-sheet to an expression or file you cannot modify it in the timeline. 
Or ideally if you do, generate a new keyframe for x-sheet similar to what happens with the other columns.

**Comment by b4zz4** [comment]
Maybe it's simpler to do it as an effect, a kind of x-sheet control.
I did something similar using transparency but it can only be seen when rendering and it's a bit tricky to use.

![2020-08-11-163558_486x269_scrot](https://user-images.githubusercontent.com/643833/89941330-3fab1780-dbf1-11ea-9eb5-d249bdab71a7.png)
![2020-08-11-163607_639x469_scrot](https://user-images.githubusercontent.com/643833/89941341-4174db00-dbf1-11ea-8990-333db69f8de5.png)

**Comment by WolfInABowl** [comment]
Speaking of an FX-based workaround, here's a plugin I wrote a little while ago that might be useful:

[WolfInABowl_Selector.zip](https://github.com/opentoonz/opentoonz/files/5059395/WolfInABowl_Selector.zip)
It takes up to 10 inputs (if you want more I can make a bigger one), you can animate the `Index` parameter to select which of the inputs is shown. If there is nothing connected to the current index, nothing will be shown.

**Comment by b4zz4** [comment]
@WolfInABowl thanks,
do you know how to compile it for gnu / linux? I managed to compile some but it does not take them

**Comment by b4zz4** [comment]
In toonz script there is this:

`scene.setCell(1, 0, N)`

Is there a way to run a script on every render frame?
Or mix it with the script that runs inside the program?

https://opentoonz.readthedocs.io/en/latest/toonzscript.html?highlight=script#scene

**Comment by gab3d** [comment]
This is both a very god idea, and a potentially powerful feature to have, too.

I think there should be some way to access the exposed frame of a column in the Function Editor, the problem could be that for each exposed drawing there should also be a reference to what level it belongs to, so there should be 2 columns for this to work, not just 1.

The great thing about this (should it be possible) is that it should be possible to dynamically populate an Xsheet column with drawings from different levels (via a File Interpolation, for example); or to control the exposed drawings with an Expression (or any other Interpolation method, really) ...the current method of setting the exposed frame could be thought of as a Stepped interpolation, where the previous frame holds until a new frame in keyed in. 

Possibilities would be broad... also possible error situations should be carefully taken into consideration, for the software not to "explode" if something goes wrong with level naming or frame numbering.

**Comment by gab3d** [comment]
@WolfInABowl 
that sonds great! :rocket: 

could this effect be included in OT standard? 
I don't know if license permits, just asking...

**Comment by b4zz4** [comment]
@gab3d 
I am recognizing emotions and positions of a face in a video to make a digital rotoscope and I could not believe that the only way to modify the spreadsheet is with lipsync.
But this has much more power it can be used for thousands of things like harassing them to a rotation, linking x-sheet, etc.
